### ISSUE TITLE
Admin login form potentially redirects to wrong URL

### ISSUE THREAD

**Issue by Manko10** [GetSimpleCMS/GetSimpleCMS#1107]
The admin login from in index.php looks like this:

    <form class="login" action="<?php echo myself(false).'?'. htmlentities($_SERVER['QUERY_STRING'], ENT_QUOTES); ?>" method="post">

which means it will effectively write `"admin?"` into the action attribute when you request the page without any parameters. This only works if the URL is exactly http://example.com/admin, but not http://example.com/admin/.

However, admin may often rewrite to admin/ due to .htaccess redirects. This results in the form being submitted to http://example.com/admin/admin? which results in a 404 response.

I'd suggest prepending `$SITEURL` to `myself()` for fixing this issue.

**Comment by tablatronix** [comment]
It uses PHP_SELF, which is usually `index.php` so I am not sure how you are getting `admin`, maybe some funky (read bad) rewrite you are using ?

( this is fixed in 3.4 because we now use scriptname NOT PHP_SELF )

Care to provide more information on how you are reproducing this ?

**Comment by Manko10** [comment]
I just had a closer look at the source code. myself() is actually defined as follows:

```
function myself($echo=true) {
	if ($echo) {
		echo htmlentities(basename($_SERVER['PHP_SELF']), ENT_QUOTES);
	} else {
		return htmlentities(basename($_SERVER['PHP_SELF']), ENT_QUOTES);
	}
}
```

So in theory it should return index.php, but it doesn't. I printed PHP_SELF to the screen and it contains the full request URI, including protocol prefix and hostname. Not sure why. Maybe a PHP update? I'm not really in the loop anymore in terms of recent PHP releases. :-)

**Comment by tablatronix** [comment]
I think this is why it was changed in 3.4 there was some apache inconsistencies.
if you server is not returning proper PHP_SELF then there are other issues you are gonna have I imagine. Also PHP self can be spoofed from client, so its possible your browser or a proxy might mess with it also.

the easiest fix it to just remove it, form actions will submit to themselves and are not even necessary.

#941

**Comment by tablatronix** [comment]
#996 

you using a reverse proxy by chance ?
What httpd ARE you using ?

**Comment by tablatronix** [comment]
#462

**Comment by Manko10** [comment]
Apache 2.4.17, php 5.6.12 with mod_fcgid.
SCRIPT_NAME also contains the URL for some reason.

**Comment by Manko10** [comment]
SCRIPT_FILENAME contains the absolute internal file path.

**Comment by tablatronix** [comment]
```php
$properties = array('SCRIPT_FILENAME', 'SCRIPT_NAME', 'PHP_SELF', 'REQUEST_URI');
printf("% 15s: %s<bR>\n", '__FILE__', __FILE__);
foreach($properties as $property){
     printf('% 15s: %s', $property, $_SERVER[$property]."<bR>\n");
}

die();
```

```php
__FILE__: C:\bin\web\dev\getsimple\develop\gsconfig.php
SCRIPT_FILENAME: C:/bin/web/dev/getsimple/develop/admin/index.php
SCRIPT_NAME: /dev/getsimple/develop/admin/index.php
PHP_SELF: /dev/getsimple/develop/admin/index.php
REQUEST_URI: /dev/getsimple/develop/admin/index.php?redirect=edit.php?id=index_custom
```

**Comment by Manko10** [comment]
__FILE__: /srv/http/virtual/janek/htdocs/test.php
SCRIPT_FILENAME: /srv/http/virtual/janek/htdocs/test.php
SCRIPT_NAME: http://janek.local/test.php
PHP_SELF: http://janek.local/test.php
REQUEST_URI: /test.php

**Comment by tablatronix** [comment]
Truly bizarre, ill look into what can cause this and how to make it bulletproof.

**Comment by tablatronix** [comment]
Using any symblinks in your path?

**Comment by Manko10** [comment]
/srv/http/virtual/janek/htdocs/test.php is a real path.

**Comment by cnb** [comment]
You have to insert that code on top of gsconfig.php

**Comment by Manko10** [comment]
That wouldn't make any difference. The GS installation is just another folder under the same document root (no symlinks either). In fact, when I first tested this, I put a print_r($_SERVER) inside the GS code with the same results.

**Comment by tablatronix** [comment]
```
SCRIPT_NAME: http://janek.local/test.php
PHP_SELF: http://janek.local/test.php
```

should not be urls

So the question how the hell are they, the server values are populated by apache, which generates them from path_info i think. So thats why i asked about symblinks, both apache and nginx have various known issues with symblink paths, but they usually affect other things, not these variables.

So I am thinking it is something specific to this apache instance, or the virtual host config you are using or how fastcgi is configured.

At first i thought it was a missing
`DirectoryIndex index.php`
or something simple.

**Comment by Manko10** [comment]
I honestly have no idea why PHP_SELF is a URL all of a sudden.

**Comment by tablatronix** [comment]
possible cause?
http://php.net/manual/en/ini.core.php#ini.cgi.fix-pathinfo

**Comment by Manko10** [comment]
That is actually part of the problem. When I set it to 1 (which seems to be the default), PHP_SELF and SCRIPT_NAME contain a URL. When I set it to 0, it looks like this:

```
__FILE__: /srv/http/virtual/janek/htdocs/test.php
SCRIPT_FILENAME: /srv/http/virtual/janek/fcgi-bin/php5.fcgi
SCRIPT_NAME: /fcgi-bin/php5.fcgi
PHP_SELF: /test.php
REQUEST_URI: /test.php
```

So PHP_SELF actually contains the file name, but SCRIPT_NAME is garbage, because it includes the FCGI handler path which is configured as follows:

```
# Define the "php-fastcgi" action
Action php-fcgid /fcgi-bin/php5.fcgi

# Map the .php and .php5 extensions to the php-fastcgi action
AddHandler php-fcgid .php
AddHandler php-fcgid .php5
```

**Comment by tablatronix** [comment]
I think there are multiple settings for these, to fix them up.
Or it cannot resolve because a path is wrong, etc.

**Comment by Manko10** [comment]
/fcgi-bin is actually located outside document root. It's just a ScriptAlias:

```
ScriptAlias /fcgi-bin "/srv/http/virtual/janek/fcgi-bin"
```

DocumentRoot is /srv/http/virtual/janek/fcgi-bin/htdocs, further virtual DocumentRoots are located under /srv/http/virtual/janek/fcgi-bin/v/$HOSTNAME

**Comment by tablatronix** [comment]
yeah this will probably need further research , not by me, ill find the most stable way to fix this issue in GS so that it can deal with these problems, or just use the full urls, I forget the reason against that but there was one.

**Comment by Manko10** [comment]
Maybe the reason HTTPS? That's a general issue in GS. I can only set HTTP or HTTPS as base URL in the settings. I cannot, however, say //example.com which would translate to whatever protocol I used to access the website.

**Comment by tablatronix** [comment]
I doubt it, but that is also a problem, and also resolved in 3.4 ( not very tested )
Feel free to try out the master branch and see if any of your problems are resolved.
